api/base.proto

```diff
@@ -33,6 +33,8 @@ message Node {
   // This is motivated by informing a management server during canary which
   // version of Envoy is being tested in a heterogeneous fleet.
   string build_version = 5;
+  // The optional version of the service.
+  string version = 6;
```

Comment:
This makes me a little uncomfortable because it's not necessarily true that there is a single service being fronted by Envoy. There could be many services. I'm wondering if there is a better way of expressing what you are trying to do here? I suppose at minimum there could be a list of service name/version pairs, but what if the service gets deployed?

Reply:
This can already all be expressed in the `metadata` field. We should add stuff to `Node` if it's a first-class concept in Envoy, are we planning on adding a notion in the CLI or bootstrap to express "service version"?

Reply:
@mattklein123 Possibly a better way to express it is the cluster version? The main reason I used to term service was because of the `--service-cluster` CLI option.

@htuch Yes there is currently a PR with `--service-version` CLI option

So if it makes more sense, I'm happy to change to refer to cluster version?

Reply:
Can you provide more context? The service-cluster is already encoding the service version as well right? (e.g., productpage-v1, reviews-v2, etc..). Why add a separate version field? And I agree with @mattklein123 that people deploy envoys in front of multiple services. In such cases, it makes sense to identify the envoy (or an autoscaled group of envoys), rather than the service(s) it is fronting.

Reply:
No, the service-cluster is using the metadata label 'app', as shown in this image from istio:

https://user-images.githubusercontent.com/164562/30971798-e6ab35fe-a460-11e7-9f9d-253a2c1e72bb.png

Which I think is good, as it means there is a constant name regardless of version - but at the same time would be useful to know the version, which is the motivation for this change.

Reply:
right. What I am saying is that you can flatten the version into the service cluster value itself right? Thats what the `app` tag is doing effectively..

Reply:
No convention AFAIK. The service cluster is something independent of applications. If you have an autoscaled group of envoys, you name that group and assign that group name to the service cluster (and possibly assign the node name to --service-node). The idea is that you trace calls across Envoys in the mesh and not the apps that Envoy is fronting.

If it so happens that there is a one to one relationship between an Envoy instance and the app that it is fronting (as is the case in Istio, where envoy runs as sidecar in pod), then well and good. But there are other scenarios where you deploy Envoy as just an edge proxy or some intermediate proxy tier, to take advantage of non-routing features. In such cases, you name that envoy cluster appropriately.

To be honest, the use of `istio-proxy` as the universal cluster name in Istio was more of a bug on our part (because kubernetes pods can be part of multiple services :( ). So the app label idea/fix you had worked out perfectly as we now trace calls across deployments (aka service versions), and not services (which can have multiple deployments). Make sense?

Reply:
The service cluster set on the command line is used in a few different somewhat random ways today. At Lyft we use it basically to identify the EC2 ASG in which the Envoy is running. There is some unfortunate history here around this CLI value and then the "local cluster" used for locality aware routing.

Per @rshriram, stepping back, what are you trying to accomplish from the end user perspective?

Reply:
At the moment tracing information is being reported in the context of the 'unit' defined by the `service-cluster` parameter. So if we ignore what that cluster represents - one or multiple services/apps - from a tracing perspective it is one named unit.

In some case (i.e. istio) the units also optionally have a defined version. From a tracing perspective, it would be useful to know what version of the unit was being used - so having the version reported as part of the tracing information.

So for example, being able to view all traces related to the 'ratings' unit (regardless of version) - but at the same time being able to do sub-queries related to version 'v1' of the 'ratings' unit.

Reply:
I guess from this perspective as long as we are clear on the global nature of the version, I don't have a problem with this. Thoughts from others?

Reply:
We can set node ID on the command line: https://envoyproxy.github.io/envoy/operations/cli.html#cmdoption-service-node (and do already trace this IIRC). I think the desire is to have a version of a cluster that is orthogonal to a node.

Reply:
Sure, but why can't this be arbitrary metadata. It seems you might want to have other IDs and annotations you want to put on traces?

Reply:
Oh sorry I forgot we add metadata to Note. You said ID and I got confused. Yes, +1, let's allow setting metadata on the CLI and then tracing that. Great idea!

Reply:
👍 sounds like a better idea. I'll close this PR, create an issue to outline the idea and then hopefully submit something next week.